0.Exploratory_Data_Analysis/nbconverted/Exploratory_Data_Analysis.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Exploratory Data Analysis on Wave Data
+
+# ### Wave1 dilated 50 data
+# Here we understand more about the data and how it is distrubuted.
+# Wave1 data are extracted features from raw images.
+# These images were processed via Cellprofiler pipelines
+#
+# Specifically wave1 is looking at Gasdermin-D and Nuclei Staining from a cell painting experiment.
+#
+# Further, nuclei were dilated using multiple values of pixel dilation. Here we use data for the 50 pixel dialation
+
+#
+# ### Wave3
+# Here we look into how the data are distributed in wave3, the same plate data as wave1 but with full cell painting dataset including the Gasdermin D channel for a total of 6 channels to analyze.
+# These data sets contain extracted features from raw image dataset. See README for information as to how raw image data was processed.
+
+# In[1]:
+
+
+import sys
+
+import matplotlib.pyplot as plt
+import pandas as pd
+import seaborn as sns
+
+get_ipython().run_line_magic("matplotlib", "inline")
+import itertools
+
+from sklearn.cluster import KMeans
+from sklearn.manifold import TSNE
+
+sys.path.append("..")
+from ..utils.utils import *
+
+# In[2]:
+
+
+# function for EDA of each input feature extraction file:
+def EDA_run(df, wave_label):
+    """Function runs EDA on imported data frame
+
+    Parameters:
+    df: pandas dataframe
+    wave_label: string of label for outputed graphs
+
+    Return:
+    graphs, and df stats
+
+    """
+    # Call function to display df shape and # of replicates present
+    df_stats(df)
+    # Drop na and reindex accordingly
+    df = df.dropna()
+    df.reindex()
+    # Check for Nans again
+    df_stats(df)
+    # Understand categorical data such as treatment and dosing
+    df[["Metadata_treatment", "Metadata_dose"]].drop_duplicates()
+    # create a list with only columns from the data frame that start with "Metadata"
+
+    treatments = df["Metadata_treatment"].unique()
+    print(treatments)
+
+    n_clusters = len(treatments)
+    # print(n_clusters)
+    samples = str(treatments).strip("[]")
+    # print(samples)
+    new_wave_label = f"{wave_label}_treatments:{samples}"
+
+    subset_number = 1500
+    df_subset = df.sample(n=subset_number)
+    # Code snipptet for metadata extraction by Jenna Tomkinson
+    df_metadata = list(df_subset.columns[df_subset.columns.str.startswith("Metadata")])
+
+    # define which columns are data and which are descriptive
+    df_descriptive = df_subset[df_metadata]
+    df_values = df_subset.drop(columns=df_metadata)
+
+    treatment_ids = df_descriptive["Metadata_treatment"]
+    # Cluster data
+    # clustering code adapted from https://www.kaggle.com/code/aussie84/clustering-with-kmeans-pca-tsne
+
+    kmeans = KMeans(n_clusters=n_clusters)
+    clustering_ori = kmeans.fit_predict(df_values)
+
+    X = df_values
```

Comment:
This may not be an issue now but I think pandas uses "references" when assigning dataframes to different variables. How can this be dangerous? Here's an example below:

```python
df = pd.DataFrame(data)

df_2 = df # reference of df is here 

df_2 = df.append(["new", "data"]) # reference of df is also here
```
So what is happening here? references are just pointers that points to the original variable that contains the data. So in this case. `df_2` contains references to `df`, why is this a problem? Well references are mutable, meaning that any manipulations conducted in `df_2` will be reflected on the original variable `df`. This can bring a  series of unexpected results.

Another question is why does Pandas do this? It's to avoid copying dataframes. If you copy dataframes, that exponentially increases memory usage. 

Copying means reserving the same object into another location in memory. Therefore you will have two identical objects in memory. 

This is just a heads up :)  computer science lecture of the day <3

Reply:
I wasn't aware of this, thank you for taking the time to educate me. This is something that I will need to keep in mind moving forward in subsequent analysis